src/constructUrl.js

```diff
@@ -113,49 +114,176 @@ var DEFAULT_OPTIONS = Object.freeze({
+  });
 
-  return url.slice(0, -1);
+  return { client, pathComponents };
 }
 
 function buildURLPublic(src, imgixParams = {}, options = {}) {
   const { disableLibraryParam } = options;
 
   const [rawSrc, params] = extractQueryParams(src);
 
-  return constructUrl(
-    rawSrc,
-    Object.assign(
-      {},
-      params,
-      imgixParams,
-      disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }
-    )
-  );
+  return constructUrl(rawSrc, {
+    ...params,
+    ...imgixParams,
+    ...(disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }),
+  });
+}
+
+/**
+ * Build a imgix source url with parameters from a raw url
+ */
+function buildSrc({
+  src: inputSrc,
+  width,
+  height,
+  disableLibraryParam,
+  disableSrcSet,
+  imgixParams,
+  disableQualityByDPR,
+}) {
+  const [rawSrc, params] = extractQueryParams(inputSrc);
+  const fixedSize = width != null || height != null;
+
+  const srcOptions = {
+    ...params,
+    ...imgixParams,
+    ...(disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }),
+    ...(fixedSize && height ? { height } : {}),
+    ...(fixedSize && width ? { width } : {}),
+  };
+
+  if (disableSrcSet) {
+    const src = constructUrl(rawSrc, srcOptions);
+    return { src, src };
+  }
+
+  if (fixedSize) {
+    const srcSet = buildSrcSet(
+      rawSrc,
+      srcOptions,
+      {
+        disableVariableQuality: disableQualityByDPR,
+      },
+      width,
+      height
+    );
+
+    const src = constructUrl(rawSrc, srcOptions);
+
+    return { src, srcSet };
+  } else {
+    warnInvalidAspectRatio(imgixParams.ar || "", config);
+
+    const { w, h, ...urlParams } = srcOptions;
+    const src = constructUrl(rawSrc, srcOptions);
+    const srcSet = buildSrcSet(rawSrc, { ...urlParams });
+
+    return { src, srcSet };
+  }
+}
+
+function buildSrcSet(rawSrc, params = {}, options = {}, width, height) {
+  const { client, pathComponents } = extractClientAndPathComponents(rawSrc);
+  const compactedParams = compactParamKeys(params, width, height);
+  return client.buildSrcSet(pathComponents.join("/"), compactedParams, options);
+}
+
+function warnInvalidAspectRatio(aspectRatio, config) {
+  const NODE_ENV = process.env.NODE_ENV;
+  const showARWarning =
+    aspectRatio != "" && aspectRatioIsValid(aspectRatio) === false;
+  if (
+    NODE_ENV !== "production" &&
+    showARWarning &&
+    config.warnings.invalidARFormat
+  ) {
+    console.warn(
+      `[Imgix] The aspect ratio passed ("${aspectRatio}") is not in the correct format. The correct format is "W:H".`
+    );
+  }
+}
+
+/**
+ * Validates that an aspect ratio is in the format w:h. If false is returned, the aspect ratio is in the wrong format.
+ */
+function aspectRatioIsValid(aspectRatio) {
+  if (typeof aspectRatio !== "string") {
+    return false;
+  }
+
+  return /^\d+(\.\d+)?:\d+(\.\d+)?$/.test(aspectRatio);
+}
+
+function buildChildProps(obj, src, attributeConfig, width, height, refType) {
```

Comment:
I think this function relates more to the components where it is called, so at least it should be located in that file.

It seems like a bit of a smell to be passing `this` around - seems like it could go wrong in the future if this function was to incorrectly assign to `obj`. I wonder if it's better if it's possible to make this pure, or just to deal with this logic being duplicated in the separate components. It's not super complicated so I'm not sure how much benefit there is to extract it into a common function.

Reply:
Yeah, I think it smells a bit too, but I wanted to try to extract/distill some of the structure of what's going on here. I'll look into refactoring this.